Hosting Dashboard: Add Date Range Controls

coder-karen opened the pull request:
Part of DOTCOM-14159

## Proposed Changes

* This PR adds a date range control, which is currently for use on the site logs page but is added as a component.
* The date range 'button' is an ~InputControl~ Button -- design specs suggested a text control but then they were also suggesting that we should use the old implementation. However after a conversation in Slack - p1755018709178949-slack-C0982082MSR - I'm using the component available from `@automattic/ui`. 
* The component started as a simple MVP datepicker, but based on an example shared in Slack it now has similar functionality to the picker from before. Slack: p1755503128722479/1755018709.178949-slack-C0982082MSR
* Several styles have been implemented to keep everything aligned as expected - mainly related to mobile rendering.

For follow-up:
- More investigation on timezone differences. Story book additions. Additional tests.

## Why are these changes being made?

* The date range picker is needed for the site logs page, part of the Hosting Dashboard v2.

## Testing Instructions

* Open up the site logs page at `http://calypso.localhost:3000/v2/sites/your-site/logs`.
* The date range picker button should be pre-selected for the past 7 days. 
* Test changing the days, and using presets, as well as selecting custom date ranges.
* Make sure the dates in your logs are correct for the ranges given. Ideally this will be tested across several different timezones.
* Note that other functionality on the logs page still hasn't been implemented such as views and actions, so it's expected that taking actions such as trying to sort won't work.

Screenshots:

Desktop:

<img width="664" height="573" alt="Screenshot 2025-08-20 at 10 37 24" src="https://github.com/user-attachments/assets/1fba9325-4d05-4b10-a14c-db345259b5e4" />


Mobile: 

<img width="336" height="570" alt="Screenshot 2025-08-20 at 10 40 48" src="https://github.com/user-attachments/assets/ca0bb5d5-f0c1-4396-9ac7-355dd2783392" />



## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you tested accessibility for your changes? Ensure the feature remains usable with various user agents (e.g., browsers), interfaces (e.g., keyboard navigation), and assistive technologies (e.g., screen readers) (PCYsg-S3g-p2).
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
    - [ ] For UI changes, have we tested the change in various languages (for example, ES, PT, FR, or DE)? The length of text and words vary significantly between languages. 
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-aUh-p2)?

coder-karen committed:
Add daterangecontrols for the site logs within hosting dashboard

coder-karen committed:
DateRange picker complete refactor, based on an example shared in Slack - see PR description for more

coder-karen committed:
Add spacing and function comment to formatLabel function

coder-karen committed:
Merge branch 'trunk' into add/hosting-dashboard-date-range-controls

coder-karen committed:
Fix merge conflict - fix siteRoute import location

mirka commented on client/dashboard/components/date-range-picker/index.tsx:
```diff
@@ -0,0 +1,372 @@
+						<div className="daterange-input__toggle">
+							<InputControl
+								value={ label }
+								onChange={ () => {} }
```
This implementation is not accessible for a few reasons:

- It's not accessibly labeled (a `Tooltip` only adds visual information).
- A screen reader user will not be able to tell that the input has button-like handlers.

The first is easily fixable but the second is more of a fundamental problem. It should semantically be a `button`. Any reason we're bending over backwards to use an `input` element? It doesn't quite make sense to me even from a visual standpoint, since the text is not editable.

coder-karen replied:
Thanks for digging in to this, and for these suggestions.

No, no reason for it to specifically be `InputControl`, other than it made sense to me at the time. The design specs suggested a `TextControl` which made less sense, but to me the logical next step was an `InputControl`. But `button` does make a lot more sense - I've updated this.

mirka commented on client/dashboard/components/date-range-picker/index.tsx:
```diff
@@ -0,0 +1,372 @@
+											style={ { minWidth: 140, flex: '1 1 0' } }
+										/>
+										<TextControl
+											placeholder="YYYY-MM-DD"
```
How about we use a `type="date"`? The hardcoded `"YYYY-MM-DD"` is not great for i18n either.

coder-karen replied:
Thanks for the suggestion -- updated!

p-jackson replied:
Using `"date"` does have the interesting effect of meaning there are now two calendars for controlling the same thing. On one hand maybe there's no harm in allowing two methods to edit the same thing. On the other hand I do think it looks a little confusing.

![CleanShot 2025-08-21 at 14.57.20@2x.png](https://app.graphite.dev/user-attachments/assets/7445478f-2786-4661-9009-a96ca11d8d91.png)

Something that we can iterate on though.

coder-karen replied:
Yes I noticed that too. I wonder if it's still needed in terms of accessibility or if it can be at least visually hidden - that said it does function. I'll look into this more for the later follow-up though.

ciampo commented:
Some additional feedback on top of @mirka 's:

- The list of presets should be given some additional semantics — for example, `listbox` + `option`
- It looks like the code could be tidied up / modularized. For example, the from/to date inputs could be extracted instead of being repeated twice for small/large viewport layouts. Also the `DropdownMenu`'s `renderContent` prop could be broken down so that fewer variables / components are declared each time the component re-renders.

Overall, I recommend taking a look at [this Storybook example](https://github.com/Automattic/wp-calypso/blob/b8eac07b35d3af60aa09b056c7cc31f6a1012daa/packages/ui/src/calendar/date-range-calendar/index.stories.tsx#L324) which already addresses many items of feedback left so far (ie. using a button for the dropdown trigger, using `<InputControl type="date" />` for the date inputs, using listbox semantics for the range presets, automatically highlighting a preset if the user manually selects the corresponding dates, etc etc).

coder-karen committed:
Change calendaar InputControl to Button

coder-karen committed:
Add the date type to the TextControl instances, removing the placeholder

p-jackson commented on client/dashboard/utils/datetime.ts:
```diff
@@ -238,3 +242,118 @@ export function getUtcOffsetDisplay( offsetHours: number ): string {
+/**
+ * Convert a date to the start of the local day.
+ */
+function toStartOfDayLocal( date: Date ): Date {
```
I think we could replace this with `date-fns` `startOfDay()` function?

coder-karen replied:
Added, thanks! The functionality works as expected with that.

p-jackson commented on client/dashboard/utils/datetime.ts:
```diff
@@ -238,3 +242,118 @@ export function getUtcOffsetDisplay( offsetHours: number ): string {
+/**
+ * Convert a date to the end of the local day.
+ */
+function toEndOfDayLocal( date: Date ): Date {
```
Same here with `date-fns` `endOfDay()`

Similar thing with the parsing functions below.

p-jackson commented on client/dashboard/utils/datetime.ts:
```diff
@@ -238,3 +242,118 @@ export function getUtcOffsetDisplay( offsetHours: number ): string {
+			'M j, Y',
+			end,
+			timezoneString
+		) }`;
```
I think this one needs to use a `sprintf()` so that the translator has more control over the sentence structure e.g. `__( '%s to %s' )`

Also this `formatLabel()` function seems sorta specific to the logging interface so I wonder if it belongs there rather than in this utility file 🤔 Maybe the same thing with the `buildTimeRangeInSeconds()` function? Seems very specific to calling the logging API.

coder-karen replied:
Thanks for the suggestions here. I moved `formatLabel` to a utils file within the picker itself rather than datetime, as I see it as relevant to the label itself in the picker rather than specifically logs focussed.

`buildTimeRangeInSeconds` moved to a logs util though.

p-jackson commented on client/dashboard/sites/logs/index.tsx:
```diff
@@ -71,6 +74,19 @@ function SiteLogs( { logType }: { logType: LogType } ) {
 	const { data: site } = useSuspenseQuery( siteBySlugQuery( siteSlug ) );
 
+	// @todo: We'll be able to remove the fallbacks once the temporary data (fields, views, actions) are removed and this component is cleaned up, as we'll return earlier if site doesn't exist.
+	const siteId = site?.ID ?? null;
```
I know you noted this as something to clean up, but just an FYI we can already rely on `site.ID` being defined since `useSuspenseQuery` will always return a `Site`, not an `undefined`.

coder-karen replied:
Yes good point, `useSuspenseQuery` came in later but I hadn't updated this specific variable yet -- updated it here anyway now.

p-jackson commented on client/dashboard/sites/logs/index.tsx:
```diff
@@ -71,6 +74,19 @@ function SiteLogs( { logType }: { logType: LogType } ) {
+	const { data } = useSuspenseQuery( {
+		...siteSettingsQuery( siteId ),
+		select: ( s ) => ( {
+			gmtOffset: s?.gmt_offset ?? 0,
+			timezoneString: s?.timezone_string ?? '',
```
It seems unfortunate that sometimes we have to use `gmtOffset` and sometimes `timezoneString` 😬 Do you have any info on when one is defined over the other? Maybe I'm misreading the code, but I'm interpreting it as it will either be one or the other.

coder-karen replied:
I'm glad you asked! I've got a follow-up task for the calendar to add various improvements (including for eg. storybook, additional tests beyond those in datetime), and I added a note to look into the timezone functionality more.
It was a head scratcher for me, and as best I can tell from the research I've done `gmt_offset` will always exist (defaults to 0 if not set), but `timezone_string` can be empty when the site [uses a numeric offset](https://developer.wordpress.org/reference/functions/wp_timezone_string/).
`date_i18n` apparently has issues with certain timezone settings such as numerical ones, which means that using `timezone_string` where possible should fix issues in those cases (eg. DST) -  IANA zones (e.g., Europe/London) apparently embed DST transition rules. 
As I understand, using 'TZDate' as suggested in the calendar package code doesn't help with this case anyway, though I've added it where possible.
For now I think some good testing across different timezones would be helpful, for the 'MVP' anyway.

p-jackson replied:
Wow, today I learned how wild the return value of `wp_timezone_string()` is!

p-jackson commented on client/dashboard/sites/logs/dataviews/fields.tsx:
```diff
@@ -44,25 +44,37 @@ const getLabelRenderer = ( renderer: string ) => {
+
+	if ( timezoneString ) {
+		/* Translators: %s is the site's timezone string */
+		dateTimeLabel = sprintf( __( 'Date & time (%s)' ), timezoneString );
```
I don't think an identical string can have different translator comments. I think it will need to be a single string with a comment "%s is the site's timezone string or UTC offset"

p-jackson commented on client/dashboard/sites/logs/dataviews/fields.tsx:
```diff
@@ -44,25 +44,37 @@ const getLabelRenderer = ( renderer: string ) => {
+	logType,
+	timezoneString,
+	gmtOffset,
+}: UseFieldsArgs ): Field< PHPLog | ServerLog >[] {
```
This isn't a blocking comment for this PR, but just something I noticed about how the types are working for these data views.

There is a lot of type casting inside this `useFields()` hook, because the items have to be told whether they are a `PHPLog` or `ServerLog`. I think the underlying issue is that this the return type should actually be `Field< PHPLog > | Field< ServerLog>`.

If you went through and changed the field arrays so that they were like:

```ts
if ( logType === LogType.PHP ) {
   return [
      // ...
   ] satisfies Field< PHPLog >;
}
```

Then the type casting wouldn't be necessary since it knows it will only ever be _either_ a php or server log, never both.

This does have some flow on issues that would need to be fixed up too though, so just leaving it as an observation for the log dataview in general. Because we would run into issues where this field type isn't compatible with a `DataView< PHPLog | ServerLog >`. But that's not really the type of the data view we want anyway. We're really rendering a `DataView< PHPLog > | DataView< ServerLog >`.

coder-karen replied:
Thanks for mentioning this. I've added a follow-up task to cover this, at DOTCOM-14273.

p-jackson commented on client/dashboard/components/date-range-picker/index.tsx:
```diff
@@ -0,0 +1,372 @@
+	useEffect( () => {
+		setToDraft( end );
+		setToStr( formatYmd( end, timezoneString, gmtOffset ) );
+	}, [ gmtOffset, end, timezoneString ] );
```
I've always found that trying to keep internal state in sync with props is error prone. Check out [You Might Not Need an Effect](https://react.dev/learn/you-might-not-need-an-effect) from the React docs.

In this case it seems the intention is to wipe out any internal state if the props are ever changed. Which seems like we could use [this pattern](https://react.dev/learn/you-might-not-need-an-effect#resetting-all-state-when-a-prop-changes) or [this pattern](https://react.dev/learn/you-might-not-need-an-effect#adjusting-some-state-when-a-prop-changes) instead.

coder-karen replied:
Thanks for the doc links and explanation on this! I've refactored this section of code now, so this should be [resetting all state](https://react.dev/learn/you-might-not-need-an-effect#resetting-all-state-when-a-prop-changes) when a prop changes.

p-jackson replied:
I am on a personal mission to share that React doc with as many people as possible 😆

p-jackson reviewed (commented):
Thank you for working on this! I haven't tested the UI yet, but leaving some code-related comments.

coder-karen committed:
Re-organizing date range component - using helpers, other components, different child components, based on feedback

coder-karen committed:
Date and timezone updates

coder-karen committed:
Avoid using useEffect to adjust state when props change

coder-karen committed:
Update datetime tests now that the datetime file has been updated

coder-karen committed:
Use TZDate when working with timezones, wherever possible

coder-karen commented:
@mirka @ciampo @p-jackson thanks so much for the reviews here.

I should have addressed all the feedback with the new commits (or with comments to suggestions) at this stage.

I also have a follow-on task for more refinements with this Date Range Picker - including a story book example, more tests, and more research into the date and time functionality itself. Initially this was to be an MVP with the minimal date range picker functionality as needed, for follow up later, but the example shared in Slack inspired me to do as much as possible for the MVP so it has become a lot more complex than I anticipated. For new feedback it would be good to see what can be included in a follow-up, or what we should fix for an MVP. Thanks again for the deep dives!

p-jackson reviewed (approved):
Thanks for the improvements. Seems like a good place to merge and keep iterating on the logging interface IMO

arthur791004 commented on client/dashboard/components/date-range-picker/utils.ts:
```diff
@@ -0,0 +1,187 @@
+import { formatDateWithOffset } from '../../utils/datetime';
+
+// Normalize/comparison helpers
+const startOfDay = ( date?: Date ) =>
```
It seems we can consider to use date-fns for these helpers 🙂

coder-karen replied:
Yes you're right, the same suggestion was made for `utils/datetime.ts` but could have been applied here as well - I've made those changes now.

arthur791004 commented on client/dashboard/components/date-range-picker/index.tsx:
```diff
@@ -0,0 +1,160 @@
+	gmtOffset,
+	timezoneString,
+}: DateRangePickerProps ) {
+	const locale = useLocale();
```
As a component, it would be better not to rely on this hook. Instead, the component should accept locale through props.

coder-karen replied:
Gotcha, the new commit should have addressed that now.

arthur791004 commented on client/dashboard/components/date-range-picker/index.tsx:
```diff
@@ -0,0 +1,160 @@
+	const desktopLabelId = `presets-label-${ instanceId }-desktop`;
+
+	const label = useMemo(
+		() => formatLabel( start, end, locale, timezoneString, gmtOffset ),
```
This doesn't seem to be a heavy computation. Maybe we don't need to memo it 🤔

coder-karen replied:
Removed it (and same with the below suggestion) for now but we can always revisit :) I'll be doing a follow-up at some stage with any potential improvements that aren't blockers just now. Though I know this wouldn't have been, I included it in this sweep to try and get things in as good a shape as possible.

arthur791004 commented on client/dashboard/components/date-range-picker/index.tsx:
```diff
@@ -0,0 +1,160 @@
+	);
+
+	// Reset internal draft state when key inputs change by remounting the inner component
+	const resetKey = useMemo(
```
Same

arthur791004 commented on client/dashboard/components/date-range-picker/date-range-content.tsx:
```diff
@@ -0,0 +1,218 @@
+	compositeActiveId: string | null;
+	setCompositeActiveId: ( id: string | null ) => void;
+	siteToday: Date;
+	siteTodayStr: string;
```
It seems weird that there is `site` concept here. Would it be better to call it `today`?

coder-karen replied:
Gave this a tidy up as well - the reasoning was that the 'today' was based on the site time, so it only really makes sense if you're trying to work out what today is based on. But that should hopefully be clear through other functions.

arthur791004 replied:
The idea is that this should be a `DateRangePicker` rather than a `SiteDateRangePicker`, so we shouldn’t include anything related to “site.” The same applies to the siteToday variable in `DateRangePickerInner`. I think both `timezoneString` and `gmtOffset` can be used to determine today based on the site. Does it make sense?

coder-karen replied:
Yes, that makes sense. We do use the same wording in the existing Date Range component too which also influenced me. I've added another commit to remove any of the remaining wording in the new picker itself (in the main component).

arthur791004 commented on client/dashboard/components/date-range-picker/date-range-content.tsx:
```diff
@@ -0,0 +1,218 @@
+						numberOfMonths={ isSmall ? 1 : 2 }
+						defaultMonth={ defaultMonth }
+						endMonth={ endMonth }
+						disabled={ { after: today } }
```
Why do we use `today` here instead of `siteToday`?

coder-karen replied:
Good point, that was redundant. Also cleaned up.

arthur791004 commented on client/dashboard/components/date-range-picker/index.tsx:
```diff
@@ -0,0 +1,160 @@
+	gmtOffset?: number;
+};
+
+const today = new Date();
```
Do we need this?

coder-karen replied:
As above :)

arthur791004 commented on client/dashboard/components/date-range-picker/style.scss:
```diff
@@ -0,0 +1,31 @@
+    justify-content: center;
+    width: 100%;
+  }
+}
```
We need empty line at the end of file

arthur791004 commented on client/dashboard/components/date-range-picker/index.tsx:
```diff
@@ -0,0 +1,160 @@
+								<span aria-hidden="true" className="daterange-input__text">
+									{ label }
+								</span>
+								<Icon icon={ calendar } size={ 24 } style={ { marginRight: 8, paddingLeft: 4 } } />
```
Can we provide the icon to button and set the position to right 🤔

coder-karen replied:
Ah of course, good catch. That's been updated now.

arthur791004 commented on client/dashboard/components/date-range-picker/date-range-content.tsx:
```diff
@@ -0,0 +1,218 @@
+						style={ { marginBottom: 16 } }
+					>
+						<VisuallyHidden id={ mobileLabelId }>{ __( 'Date range presets' ) }</VisuallyHidden>
+						<PresetsListbox
```
It looks weird on specific size

<img width="584" height="807" alt="image" src="https://github.com/user-attachments/assets/93bd493b-5aad-44fc-9cc7-b26473b8fed4" />

coder-karen replied:
Yes good point, on a very specific screen size it does look off. I made some tweaks, though I still don't think it's ideal (the single calendar view seems a little janky in both placement and when clicking to an earlier month, when the new month displays). I think I'll dig into this more for a follow-up.

arthur791004 commented on client/dashboard/components/date-range-picker/date-range-content.tsx:
```diff
@@ -0,0 +1,218 @@
+						as="div"
+						spacing={ 1 }
+						className="daterange-presets"
+						style={ { marginBottom: 16 } }
```
Instead of using marginBottom, would it be better to add another VStack and set the spacing 🤔

arthur791004 replied:
Also, can we move VStack to `PresetsListbox` itself? It seems the component needs it internally 🤔

coder-karen replied:
I hope I understood you correctly with the tweaks I made here on the new commit. Same as with the below comment. Let me know if not!

arthur791004 replied:
Yeah now `PresetsListbox` looks good to me but we need to update the usage. For example,
* Remove `<VisuallyHidden id={ mobileLabelId }>{ __( 'Date range presets' ) }</VisuallyHidden>` as it's added inside the `PresetsListbox`
* We can also move the `daterange-presets` to the component itself 🤔

coder-karen replied:
Thanks, yes, cleaned that up now. Did you mean the `daterange-presets` classname? I've moved that too.

arthur791004 commented on client/dashboard/components/date-range-picker/date-range-content.tsx:
```diff
@@ -0,0 +1,218 @@
+			: { from: fromDraft, to: toDraft };
+
+	return (
+		<div style={ { padding: 12 } }>
```
Would it be better to use a VStack here to keep the spacing consistent instead of setting margins individually? 🤔

coder-karen committed:
Addressing new feedback - component rearranging, property names, date-fns usage, display fixes

coder-karen committed:
Update siteToday wording to just be today, remove link to 'site'

coder-karen committed:
PresetsListbox cleanup

arthur791004 commented on client/dashboard/components/date-range-picker/index.tsx:
```diff
@@ -0,0 +1,162 @@
+	const label = formatLabel( start, end, locale, timezoneString, gmtOffset );
+
+	// Reset internal draft state when key inputs change by remounting the inner component
+	const resetKey = `${ formatYmd( start, timezoneString, gmtOffset ) }|${ formatYmd(
```
Not a blocker: would it be better to join the data by `|`?

```js
const resetKey = [
  formatYmd( start, timezoneString, gmtOffset ),
  formatYmd( end, timezoneString, gmtOffset ),
  timezoneString ?? '',
  gmtOffset ?? '',
].join( '|' );
```

arthur791004 commented on client/dashboard/components/date-range-picker/index.tsx:
```diff
@@ -0,0 +1,162 @@
+	const [ compositeActiveId, setCompositeActiveId ] = useState< string | null >( null );
+
+	const today = useMemo(
+		() => parseYmdLocal( formatYmd( new Date(), timezoneString, gmtOffset ) )!,
```
Do we need the ending `!`?

arthur791004 commented on client/dashboard/sites/logs/index.tsx:
```diff
@@ -168,6 +195,15 @@ function SiteLogs( { logType }: { logType: LogType } ) {
 
 	return (
 		<PageLayout header={ <PageHeader title={ __( 'Logs' ) } /> }>
+			<DateRangePicker
```
I think it would be better to move it into the main props of the `CalloutOverlay`

arthur791004 reviewed (approved):
Looks better now! We can move it forward and iterate it in the follow-up PR 🙂

coder-karen committed:
Merge remote-tracking branch 'origin/trunk' into add/hosting-dashboard-date-range-controls

a8ci18n commented:
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/19953367

Some locales (Hebrew, Japanese) have been temporarily machine-translated due to translator availability. All other translations are usually ready within a few days. Untranslated and machine-translated strings will be sent for translation next Monday and are expected to be completed by the following Friday.

Thank you @coder-karen for including a screenshot in the description! This is really helpful for our translators.
